File: library/src/main/java/app/suprsend/database/SQLDataHelper.java
package app.suprsend.database;

import java.util.ArrayList;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import app.suprsend.base.Logger;

public class SQLDataHelper extends SQLiteOpenHelper {
    private static final String DATABASE_NAME = "suprsend.db";
    private static final int DATABASE_VERSION = 1;

    // Table : events
    private static final String TABLENAME_events = "events";
    private static final String events_Id = "Id";
    private static final String events_Value = "value";
    private static final String events_IsDirty = "isDirty";
    private static final String events_TimeStamp = "timeStamp";
    private static final String events_Uuid = "uuid";

    // Table : config
    private static final String TABLENAME_config = "config";
    private static final String config_Id = "id";
    private static final String config_Key = "key_name";
    private static final String config_Value = "value";

    private final SQLiteDatabase db;

    private static final String TAG = "database_log";

    public SQLDataHelper(Context context) {
        super(context, DATABASE_NAME, null, DATABASE_VERSION);
        db = getWritableDatabase();
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        String query1 = "CREATE TABLE " + TABLENAME_events + " (" + events_Id + " INTEGER PRIMARY KEY AUTOINCREMENT," + events_Value + " TEXT ," + events_IsDirty + " INTEGER ," + events_TimeStamp + " INTEGER ," + events_Uuid + " TEXT " + ");";
        String query2 = "CREATE TABLE " + TABLENAME_config + " (" + config_Id + " INTEGER PRIMARY KEY AUTOINCREMENT," + config_Key + " TEXT ," + config_Value + " TEXT " + ");";
        try {
            db.execSQL(query1);
            db.execSQL(query2);
        } catch (Exception e) {
            Logger.INSTANCE.e("db","onCreate",e);
        }
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        try {
            db.execSQL("DROP TABLE IF EXISTS " + TABLENAME_events);
            db.execSQL("DROP TABLE IF EXISTS " + TABLENAME_config);
        } catch (Exception e) {
            Logger.INSTANCE.e("db","onUpgrade",e);
        }
        onCreate(db);
    }

    public void insert_events(Event_Model table_model_obj) {
        ContentValues contentValues = new ContentValues();
        contentValues.put(events_Value, table_model_obj.getValue());
        contentValues.put(events_IsDirty, table_model_obj.isDirty());
        contentValues.put(events_TimeStamp, table_model_obj.getTimeStamp());
        contentValues.put(events_Uuid, table_model_obj.getUuid());
        try {
            db.insert(TABLENAME_events, null, contentValues);
        } catch (Exception e) {
            Logger.INSTANCE.e("db","insert_events",e);
        }
    }

    public ArrayList<Event_Model> geteventsList(int isDirty, long limit) {
        ArrayList<Event_Model> list = new ArrayList<Event_Model>();
        Cursor c = null;
        try {
            String query = "select * from " + TABLENAME_events + " where " + events_IsDirty + " = " + isDirty + " ORDER BY " + events_TimeStamp + " LIMIT " + limit;
            c = db.rawQuery(query, null);
            if (c != null && c.moveToNext()) {
                do {
                    Event_Model log = new Event_Model();
                    log.setId(c.getLong(c.getColumnIndexOrThrow(events_Id)));
                    log.setValue(c.getString(c.getColumnIndexOrThrow(events_Value)));
                    log.setDirty(c.getInt(c.getColumnIndexOrThrow(events_IsDirty)));
                    log.setTimeStamp(c.getInt(c.getColumnIndexOrThrow(events_TimeStamp)));
                    list.add(log);
                } while (c.moveToNext());
            }
            c.close();
        } catch (Exception e) {
            Logger.INSTANCE.e("db","geteventsList",e);
        } finally {
            if (c != null)
                c.close();
        }
        return list;
    }

    private void insert_config(Config_Model table_model_obj) {
        ContentValues contentValues = new ContentValues();
        contentValues.put(config_Key, table_model_obj.getKey());
        contentValues.put(config_Value, table_model_obj.getValue());
        try {
            db.insert(TABLENAME_config, null, contentValues);
        } catch (Exception e) {
            Logger.INSTANCE.e("db","insert_config",e);
        }
    }

    public void insert_configByKey(Config_Model table_model_obj) {
        Cursor c = null;
        try {
            String q1 = " SELECT * FROM " + TABLENAME_config + " where " + config_Key + " ='" + table_model_obj.getKey() + "'";

            c = db.rawQuery(q1, null);
            if (c != null && c.getCount() > 0) {
                if (c.moveToFirst()) {
                    ContentValues contentValues = new ContentValues();
                    contentValues.put(config_Value, table_model_obj.getValue());
                    try {
                        db.update(TABLENAME_config, contentValues, config_Key + "= ? ", new String[]{table_model_obj.getKey()});
                    } catch (Exception e) {
                        Logger.INSTANCE.e("db","insert_configByKey",e);
                    }
                }
            } else {
                insert_config(table_model_obj);
            }
            c.close();
        } catch (Exception e) {
            Logger.INSTANCE.e("db","insert_configByKey",e);
        } finally {
            if (c != null)
                c.close();
        }
    }

    public void deleteeventsByID(String ids) {
        try {
            db.execSQL("DELETE FROM " + TABLENAME_events + " WHERE " + events_Id + " IN (" + ids + ")");
        } catch (Exception e) {
            Logger.INSTANCE.e("db","deleteeventsByID",e);
        }
    }

    public void deleteAllEvents() {
        try {
            db.execSQL("DELETE FROM " + TABLENAME_events);
        } catch (Exception e) {
            Logger.INSTANCE.e("db","deleteAllEvents",e);
        }
    }

    //Starts Config table
    public Config_Model getconfigByKey(String key) {
        Config_Model log = new Config_Model();
        Cursor c = null;
        try {
            String query = "select * from " + TABLENAME_config + " where " + config_Key + " ='" + key + "'";
            c = db.rawQuery(query, null);
            if (c != null && c.moveToNext()) {
                do {
                    log.setId(c.getInt(c.getColumnIndexOrThrow(config_Id)));
                    log.setKey(c.getString(c.getColumnIndexOrThrow(config_Key)));
                    log.setValue(c.getString(c.getColumnIndexOrThrow(config_Value)));

                } while (c.moveToNext());
            }
            c.close();
        } catch (Exception e) {
            Logger.INSTANCE.e("db","getconfigByKey",e);
        } finally {
            if (c != null)
                c.close();
        }
        return log;
    }

    public void deleteAllConfigs() {
        try {
            db.execSQL("DELETE FROM " + TABLENAME_config);
        } catch (Exception e) {
            Logger.INSTANCE.e("db","deleteAllConfigs",e);
        }
    }
}

File: library/src/main/java/app/suprsend/database/SQLDataHelper.kt
package app.suprsend.database

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import app.suprsend.base.Logger.e

internal class SQLDataHelper(context: Context?) : SQLiteOpenHelper(context, DATABASE_NAME, null, DATABASE_VERSION) {
    private val db: SQLiteDatabase = writableDatabase

    override fun onCreate(db: SQLiteDatabase) {
        try {
            val query1 = "CREATE TABLE $TABLENAME_events ($events_Id INTEGER PRIMARY KEY AUTOINCREMENT,$events_Value TEXT ,$events_IsDirty INTEGER ,$events_TimeStamp INTEGER ,$events_Uuid TEXT );"
            val query2 = "CREATE TABLE $TABLENAME_config ($config_Id INTEGER PRIMARY KEY AUTOINCREMENT,$config_Key TEXT ,$config_Value TEXT );"
            db.execSQL(query1)
            db.execSQL(query2)
        } catch (e: Exception) {
            e("db", "onCreate", e)
        }
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        try {
            try {
                db.execSQL("DROP TABLE IF EXISTS $TABLENAME_events")
                db.execSQL("DROP TABLE IF EXISTS $TABLENAME_config")
            } catch (e: Exception) {
                e("db", "onUpgrade", e)
            }
            onCreate(db)
        } catch (e: Exception) {
            e("db", "onUpgrade", e)
        }
    }

    fun insertEvents(table_model_obj: Event_Model) {
        try {
            val contentValues = ContentValues()
            contentValues.put(events_Value, table_model_obj.value)
            contentValues.put(events_IsDirty, table_model_obj.isDirty)
            contentValues.put(events_TimeStamp, table_model_obj.timeStamp)
            contentValues.put(events_Uuid, table_model_obj.uuid)
            db.insert(TABLENAME_events, null, contentValues)
        } catch (e: Exception) {
            e("db", "insertEvents", e)
        }
    }

    fun getEventsList(isDirty: Int, limit: Long): ArrayList<Event_Model> {
        val list = ArrayList<Event_Model>()
        var cursor: Cursor? = null
        try {
            val query = "select * from $TABLENAME_events where $events_IsDirty = $isDirty ORDER BY $events_TimeStamp LIMIT $limit"
            cursor = db.rawQuery(query, null)
            if (cursor != null && cursor.moveToNext()) {
                do {
                    val log = Event_Model()
                    log.id = cursor.getLong(cursor.getColumnIndexOrThrow(events_Id))
                    log.value = cursor.getString(cursor.getColumnIndexOrThrow(events_Value))
                    log.isDirty = cursor.getInt(cursor.getColumnIndexOrThrow(events_IsDirty))
                    log.timeStamp = cursor.getInt(cursor.getColumnIndexOrThrow(events_TimeStamp)).toLong()
                    list.add(log)
                } while (cursor.moveToNext())
            }
            cursor.close()
        } catch (e: Exception) {
            e("db", "getEventsList", e)
        } finally {
            cursor?.close()
        }
        return list
    }

    private fun insertConfig(table_model_obj: Config_Model) {
        val contentValues = ContentValues()
        contentValues.put(config_Key, table_model_obj.key)
        contentValues.put(config_Value, table_model_obj.value)
        try {
            db.insert(TABLENAME_config, null, contentValues)
        } catch (e: Exception) {
            e("db", "insertConfig", e)
        }
    }

    fun insertConfigByKey(table_model_obj: Config_Model) {
        var cursor: Cursor? = null
        try {
            val q1 = " SELECT * FROM " + TABLENAME_config + " where " + config_Key + " ='" + table_model_obj.key + "'"
            cursor = db.rawQuery(q1, null)
            if (cursor != null && cursor.count > 0) {
                if (cursor.moveToFirst()) {
                    val contentValues = ContentValues()
                    contentValues.put(config_Value, table_model_obj.value)
                    try {
                        db.update(TABLENAME_config, contentValues, "$config_Key= ? ", arrayOf(table_model_obj.key))
                    } catch (e: Exception) {
                        e("db", "insertConfigByKey", e)
                    }
                }
            } else {
                insertConfig(table_model_obj)
            }
            cursor.close()
        } catch (e: Exception) {
            e("db", "insertConfigByKey", e)
        } finally {
            cursor?.close()
        }
    }

    fun deleteEventsByID(ids: String) {
        try {
            db.execSQL("DELETE FROM $TABLENAME_events WHERE $events_Id IN ($ids)")
        } catch (e: Exception) {
            e("db", "deleteEventsByID", e)
        }
    }

    fun deleteAllEvents() {
        try {
            db.execSQL("DELETE FROM $TABLENAME_events")
        } catch (e: Exception) {
            e("db", "deleteAllEvents", e)
        }
    }

    //Starts Config table
    fun getConfigByKey(key: String): Config_Model {
        val log = Config_Model()
        var cursor: Cursor? = null
        try {
            val query = "select * from $TABLENAME_config where $config_Key ='$key'"
            cursor = db.rawQuery(query, null)
            if (cursor != null && cursor.moveToNext()) {
                do {
                    log.id = cursor.getInt(cursor.getColumnIndexOrThrow(config_Id))
                    log.key = cursor.getString(cursor.getColumnIndexOrThrow(config_Key))
                    log.value = cursor.getString(cursor.getColumnIndexOrThrow(config_Value))
                } while (cursor.moveToNext())
            }
            cursor.close()
        } catch (e: Exception) {
            e("db", "getConfigByKey", e)
        } finally {
            cursor?.close()
        }
        return log
    }

    fun deleteAllConfigs() {
        try {
            db.execSQL("DELETE FROM $TABLENAME_config")
        } catch (e: Exception) {
            e("db", "deleteAllConfigs", e)
        }
    }

    companion object {
        private const val DATABASE_NAME = "suprsend.db"
        private const val DATABASE_VERSION = 1

        // Table : events
        private const val TABLENAME_events = "events"
        private const val events_Id = "Id"
        private const val events_Value = "value"
        private const val events_IsDirty = "isDirty"
        private const val events_TimeStamp = "timeStamp"
        private const val events_Uuid = "uuid"

        // Table : config
        private const val TABLENAME_config = "config"
        private const val config_Id = "id"
        private const val config_Key = "key_name"
        private const val config_Value = "value"
    }
}